[visionOS] Under thermal throttling Safari webpage went blank when navigating

mwyrzykowski opened the pull request:
#### bd0c3f38c2e8a609fe52a4c87e2fd4ee2e6fe06b
<pre>
[visionOS] Under thermal throttling Safari webpage went blank when navigating
<a href="https://bugs.webkit.org/show_bug.cgi?id=261374">https://bugs.webkit.org/show_bug.cgi?id=261374</a>
&lt;radar://113230434&gt;

Reviewed by Dean Jackson.

When the system is throttled due to thermal pressure, we can easily hit process
timeouts and then web pages will fail to load.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::adjustedTimeoutForThermalState):
(WebKit::AuxiliaryProcessProxy::AuxiliaryProcessProxy):
Query if thermal mitigation is enabled and increase process timeout if so.

* Source/WebCore/platform/ThermalMitigationNotifier.cpp:
(WebCore::ThermalMitigationNotifier::isThermalMitigationEnabled):
Default implementation.

* Source/WebCore/platform/ThermalMitigationNotifier.h:
* Source/WebCore/platform/cocoa/ThermalMitigationNotifier.mm:
(WebCore::isThermalMitigationEnabled):
(-[WebThermalMitigationObserver thermalMitigationEnabled]):
(WebCore::ThermalMitigationNotifier::isThermalMitigationEnabled):
Add static member function to query if thermal mitigation is enabled
as we don&apos;t need an instance function in AuxiliaryProcessProxy

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Make ThermalMitigationNotifier.h a private header.

Canonical link: <a href="https://commits.webkit.org/268149@main">https://commits.webkit.org/268149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69fce90f19b1574829ae9a6e68b1e6b33fb31503

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20727 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17647 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19336 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19214 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16428 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21610 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17191 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21537 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17951 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17018 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4477 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21384 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/17787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -42,13 +42,54 @@
+    static int thermalState = kOSThermalPressureLevelNominal;
+    static dispatch_once_t once = 0L;
+
+    dispatch_once(&once, ^{
```
can you use Objective C blocks in C++ files?

mwyrzykowski replied:
Yes, it’s an llvm C extension

hortont424 commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -45,10 +45,54 @@
+
+    dispatch_once(&once, ^{
+        int localToken;
+        uint32_t result = notify_register_dispatch(kOSThermalNotificationPressureLevelName, &localToken, dispatch_get_main_queue(), ^(int token) {
```
Any reason you chose this level of thermal notification instead of NSProcessInfo like in https://commits.webkit.org/267480@main?

hortont424 commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -45,10 +45,54 @@
+
 AuxiliaryProcessProxy::AuxiliaryProcessProxy(bool alwaysRunsAtBackgroundPriority, Seconds responsivenessTimeout)
-    : m_responsivenessTimer(*this, responsivenessTimeout)
+    : m_responsivenessTimer(*this, adjustedTimeoutForThermalState(responsivenessTimeout))
```
I'm not familiar with the mechanism via which the responsiveness timer will stop pending page loads -- what happens?

mwyrzykowski replied:
Yes, I made a note in the commit message but I didn't want to make this a .mm file as it and its callers appear to be used in non-Cocoa contexts. I could create a new .mm file with a C-header and call it from here, that would also address the issue.

This code is mostly duplicated straight from NSProcessInfo however with some small customization.

hortont424 replied:
Well, I wonder if you should just use ThermalMitigationNotifier, which abstracts over the concept with a C++ API??

mwyrzykowski replied:
That sounds great, I will try that - thank you!

mwyrzykowski replied:
@hortont424 I updated the PR to use `ThermalMitigationNotifier` instead so I don't have to import `libkern/OSThermalNotifcation.h` - is this acceptable now?

mwyrzykowski committed:
[visionOS] Under thermal throttling Safari webpage went blank when navigating
https://bugs.webkit.org/show_bug.cgi?id=261374
<radar://113230434>

Reviewed by Dean Jackson.

When the system is throttled due to thermal pressure, we can easily hit process
timeouts and then web pages will fail to load.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::adjustedTimeoutForThermalState):
(WebKit::AuxiliaryProcessProxy::AuxiliaryProcessProxy):
Query if thermal mitigation is enabled and increase process timeout if so.

* Source/WebCore/platform/ThermalMitigationNotifier.cpp:
(WebCore::ThermalMitigationNotifier::isThermalMitigationEnabled):
Default implementation.

* Source/WebCore/platform/ThermalMitigationNotifier.h:
* Source/WebCore/platform/cocoa/ThermalMitigationNotifier.mm:
(WebCore::isThermalMitigationEnabled):
(-[WebThermalMitigationObserver thermalMitigationEnabled]):
(WebCore::ThermalMitigationNotifier::isThermalMitigationEnabled):
Add static member function to query if thermal mitigation is enabled
as we don't need an instance function in AuxiliaryProcessProxy

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Make ThermalMitigationNotifier.h a private header.

Canonical link: https://commits.webkit.org/268149@main
